src/answers.sql

```diff
@@ -1,15 +1,118 @@
+                ELSE m.mount
+            END final_mount
+        FROM users u
+            JOIN accounts a on u.id = a.user_id
+            JOIN movements m on a.id = m.account_from OR a.id = m.account_to
+    ) AS accounts_with_movements
+    JOIN (
+        SELECT user_id, SUM(mount) total_mount FROM accounts GROUP BY user_id
+    ) AS total_accounts_mounts ON total_accounts_mounts.user_id = account_user_id
+    GROUP BY account_user_id, user_name, total_mount
+    ORDER BY final_mount DESC
+    LIMIT 3;
 -- 5
+BEGIN TRANSACTION;
+    --- a.
+    CREATE FUNCTION get_account_final_mount(param_account_id uuid)
+        RETURNS DOUBLE PRECISION
+        LANGUAGE plpgsql
+        AS $$
+            DECLARE var_final_mount DOUBLE PRECISION;
+            BEGIN
+                SELECT (total_mount + SUM(final_mount)) final_mount INTO var_final_mount
+                    FROM (
+                        SELECT a.id account_id, u.name  user_name,
+                            CASE
+                                WHEN m.account_to = a.id THEN m.mount
+                                WHEN m.type IN ('TRANSFER', 'OUT', 'OTHER') THEN m.mount * -1
+                                ELSE m.mount
+                            END final_mount
+                        FROM users u
+                            JOIN accounts a on u.id = a.user_id
+                            JOIN movements m on a.id = m.account_from OR a.id = m.account_to
+                            WHERE a.id = param_account_id
+                    ) AS accounts_with_movements
+                        JOIN (
+                             SELECT id, SUM(mount) total_mount FROM accounts GROUP BY id
+                        ) AS total_accounts_mounts ON total_accounts_mounts.id = account_id
+                    GROUP BY account_id, user_name, total_mount
+                    ORDER BY final_mount DESC;
+                RETURN var_final_mount;
+            END
+        $$;
+    SELECT get_account_final_mount('3b79e403-c788-495a-a8ca-86ad7643afaf') user_a_final_mount,
+           get_account_final_mount('fd244313-36e5-4a17-a27c-f8265bc46590') user_b_final_mount;
+    --- b.
+    INSERT INTO movements (id, type, account_from, account_to, mount)
+        VALUES (gen_random_uuid(), 'TRANSFER', '3b79e403-c788-495a-a8ca-86ad7643afaf', 'fd244313-36e5-4a17-a27c-f8265bc46590', 50.75);
+    --- c.
+    DO $$
+        DECLARE
+            user_a_mount DOUBLE PRECISION;
+        BEGIN
+            SELECT get_account_final_mount('3b79e403-c788-495a-a8ca-86ad7643afaf') INTO user_a_mount;
+            IF user_a_mount >= 731823.56
+                THEN
+                    INSERT INTO movements (id, type, account_from, account_to, mount)
+                        VALUES (gen_random_uuid(), 'OUT', '3b79e403-c788-495a-a8ca-86ad7643afaf', null, 731823.56);
+                ELSE
+                    --- d.
+                    BEGIN
+                        RAISE EXCEPTION 'Operation failed - Insufficient amount';
```

Comment:
here you have to answer yes o not, please update it

Reply:
Done.